app/controller/button/TracingMixin.js

```diff
@@ -0,0 +1,275 @@
+                if (me.util.lineStringNotEmpty(me.tracingFeature) && !me.tracingFeatureArray.includes(feature)) {
+                    return;
+                }
+
+                hit = true;
+                var coord = me.map.getCoordinateFromPixel(event.pixel);
+
+                // second click on the tracing feature: append the ring coordinates
+                if (me.util.lineStringNotEmpty(me.tracingFeature) && me.tracingFeatureArray.includes(feature)) {
+                    me.tracingEndPoint = me.tracingFeature.getGeometry().getClosestPoint(coord);
+                    var appendCoords = me.util.getPartialSegmentCoords(
+                        me.tracingFeature,
+                        me.tracingStartPoint,
+                        me.tracingEndPoint
+                    );
+
+                    // send coordinates to parent component
+                    if (me.getView()) {
+                        me.getView().fireEvent('tracingend', appendCoords);
+                    }
+
+                    me.tracingFeature.getGeometry().setCoordinates([]);
+                    me.tracingFeatureArray = [];
+                    me.previewLine.getGeometry().setCoordinates([]);
+                }
+
+                var geom = feature.clone().getGeometry();
+                // start tracing on the feature ring
+                var coords = geom.getCoordinates();
+
+                me.tracingFeature.getGeometry().setCoordinates(
+                    coords
+                );
+                me.tracingFeatureArray.push(feature);
+                me.tracingStartPoint = me.tracingFeature.getGeometry().getClosestPoint(coord);
+            },
+            me.forEachFeatureOptions
+        );
+
+        if (!hit) {
+            // clear current tracing feature & preview
+            me.previewLine.getGeometry().setCoordinates([]);
+            me.tracingFeature.getGeometry().setCoordinates([]);
+            me.tracingFeatureArray = [];
+        }
+    },
+
+    /**
+     * Create the tracing geometry when pointer is moved.
+     *
+     * @param {Event} event The OpenLayers move event
+     */
+    onTracingPointerMove: function (event) {
+        var me = this;
+
+        if (me.util.lineStringNotEmpty(me.tracingFeature) && me.editingIsActive) {
+            var coordOnFoundFeature = null;
+            me.map.forEachFeatureAtPixel(
+                event.pixel,
+                function (foundFeature) {
+
+                    // find coordinate on found feature
+                    if (me.tracingFeatureArray.includes(foundFeature)) {
+                        coordOnFoundFeature = me.map.getCoordinateFromPixel(event.pixel);
+                    }
+
+                    // update tracing feature and array
+                    if (me.tracingFeatureArray.includes(foundFeature)) {
+
+                        // check if found feature is last of array
+                        var i = me.tracingFeatureArray.indexOf(foundFeature);
```

Comment:
It might make sense to refactor this functionality into a dedicated function, in order to improve readability and testability.

Reply:
done with 6708b7c79d11e54b6eb018c311a4784f7c40c933